### ISSUE TITLE
2D input format dependency

### ISSUE THREAD

**Issue by edrdos101** [facebookresearch/VideoPose3D#38]
Hi, could you provide any specifics around expected 2D keypoints input? Specifically, does the network expects a vector of keypoints with joints in a specific order? By looking at test data you supplied, they stay consistent - i.e. left arm (11,12,13) etc... Perhaps the input keypoints also have to be resized?

I tried feeding keypoints generated with CPN and [AlphaPose](https://github.com/MVIG-SJTU/AlphaPose) on my own dataset but seem to be getting a random set of 3D points

**Comment by dariopavllo** [comment]
Yes, the model expects the input keypoints to be in a specific order -- specifically, the order that was used when training the model. It also depends on the set of keypoints that you have (whether they are from H3.6M or COCO).

Both CPN and AlphaPose use the COCO keypoint layout, so if you want to experiment with a pretrained model, you should use one that was trained on COCO keypoints. We have one that was trained on Detectron ( https://dl.fbaipublicfiles.com/video-pose-3d/d-pt-243.bin ) and one that was trained on CPN ( https://dl.fbaipublicfiles.com/video-pose-3d/cpn-pt-243.bin ). These should be fine if you figure out the correct order.

Of course, if you want the best accuracy you should retrain the model from scratch on the new keypoints (or at least fine-tune it).

**Comment by edrdos101** [comment]
Thanks for your reply, I am trying to run COCO based AlphaPose on a cpn-pt-243 model but still get completely random 3D points (they look nothing like the actual image so it's not an issue of accuracy). I did write some custom code just to test the prediction part and go from there where I do the normalisation of keypoints to screen size, then add padding etc. as in the UnchunkedGenerator class. It all looks good up to a point where I call the model_pos which produces some randomly scatter 3D points. Interestingly, even the original 2 dimensions get thrown around randomly.

Original Image (with AlphaPose keypoints)
![_screenshot_12 04 2019](https://user-images.githubusercontent.com/46657022/56024659-652bc680-5d08-11e9-90e7-4f433c6530a2.png)

Output of the Temporal Model with cpn-pt-243 weights loaded into it
![Figure_1](https://user-images.githubusercontent.com/46657022/56024654-60671280-5d08-11e9-9114-58f5ee8a9b04.png)

**Comment by dariopavllo** [comment]
Make sure that all the data preprocessing is done correctly.

- Are your keypoints in the [correct ](https://github.com/facebookresearch/Detectron/blob/8c1e2afb74b3b5efaa8dc198b1bac81398593634/detectron/utils/keypoints.py#L30-L52)order?
- Did you normalize them properly (look at our code)? The longest side (usually the width) of your screen resolution should be mapped to [-1, 1] and everything should be zero-centered. The 2D pose should be upside down in Cartesian coordinates.
- Ideally you should create a custom dataset and use our code if you want to be sure that everything is done correctly.

**Comment by edrdos101** [comment]
I seem to be getting somewhere, looks like much better now, although here's another issue, the order of output 3D points seemed to be messed up. Would the the output retain original COCO ordering? {'nose', 'left eye', 'right eye' etc}?

**Comment by dariopavllo** [comment]
The output points follow the Human3.6M skeleton, not the COCO one (for which 3D data does not exist). In our model, the set of input and output keypoints may be different.

It's not just the order -- some keypoints are actually different. For instance, in COCO you have nose/eyes/ears which are absent in Human3.6M, but in H3.6M you have the pelvis, spine and neck.

Look at this comment for the list of Human3.6M joints: https://github.com/facebookresearch/VideoPose3D/issues/36#issuecomment-481026263

**Comment by edrdos101** [comment]
Embarrassingly I think this was my problem all along as I assumed the ordering was consistent. Thank you very much for your help, I think I can now get on with it!

Best Regards

**Comment by OctaM** [comment]
I am facing a similar issue: I use openpose to predict the keypoints and then I convert the openpose keypoints to coco format. Everything looks fine in the 2D video, but when I call model_pos I get some points but I think something is messed up and nothing shows in the 3D prediction window. I figured that if I normalize 2D predictions between 0-1 I get a pose in 3D but it's not moving along with the video. 

This is how the prediction looks after normalization of the 2D keypoints. 
![openpose_normalized_photo](https://user-images.githubusercontent.com/8879730/57179363-e13ba900-6e85-11e9-8d8c-9bd3ffc3378e.jpg)
